sklearn/utils/__init__.py

```diff
@@ -89,6 +90,9 @@
 
 def _array_indexing(array, key, key_dtype, axis):
     """Index an array or scipy.sparse consistently across NumPy version."""
+    xp, is_array_api = get_namespace(array)
+    if is_array_api and not _is_numpy_namespace(xp):
```

Comment:
I'm not sure. From messing around a bit with this what I realised is that if we want to use `if is_array_api` only here, we need to use the same `if` statement in `_array_api.py`/in the `ensure_namespace()` function. Otherwise you get errors because a Numpy ndarray is not exactly the same as a Array API array.

<details>
<summary>An example of "mismatching" if statements</summary>
```python
sklearn/utils/_param_validation.py:213: in wrapper
    return func(*args, **kwargs)
sklearn/model_selection/_split.py:2805: in train_test_split
    return list(
sklearn/model_selection/_split.py:2807: in <genexpr>
    (_safe_indexing(a, train), _safe_indexing(a, test)) for a in arrays
sklearn/utils/__init__.py:329: in _safe_indexing
    return _array_indexing(X, indices, indices_dtype, axis=axis)
sklearn/utils/__init__.py:95: in _array_indexing
    return xp.take(array, key, axis=axis)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = Array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9.],
       [10., 11., 12., 13., 14., 15., 16., 17., 18., 19.],
...1., 82., 83., 84., 85., 86., 87., 88., 89.],
       [90., 91., 92., 93., 94., 95., 96., 97., 98., 99.]], dtype=float64)
indices = array([9, 1, 6, 7, 3, 0, 5])

    def take(x: Array, indices: Array, /, *, axis: Optional[int] = None) -> Array:
        """
        Array API compatible wrapper for :py:func:`np.take <numpy.take>`.

        See its docstring for more information.
        """
        if axis is None and x.ndim != 1:
            raise ValueError("axis must be specified when ndim > 1")
        if indices.dtype not in _integer_dtypes:
            raise TypeError("Only integer dtypes are allowed in indexing")
        if indices.ndim != 1:
            raise ValueError("Only 1-dim indices array is supported")
>       return Array._new(np.take(x._array, indices._array, axis=axis))
E       AttributeError: 'numpy.ndarray' object has no attribute '_array'
```
</details>

And there are cases where we want to special case Numpy ndarrays, either because we want to preserve the old behaviour, there is an optimisation possible, etc. So with that need I wonder if we will get ourselves tangled up in when to use `if is_array_api` and when `if is_array_api and not is_numpy`. Or ending up having to modify `if` statements at a later time to make things work.

But I also dislike having to write `if is_array_api and not _is_numpy_namespace(xp):` essentially everywhere. It makes me think should `is_array_api` be `False` for a Numpy ndarray? It isn't a Array API array, so why is this variable true? But that feels like a big change to make :-/

Reply:
> Numpy ndarray is not exactly the same as a Array API array.

In the longer term, they should converge. For instance in the [numpy dev doc](https://numpy.org/devdocs/reference/array_api.html) (for the upcoming numpy 2), it is stated:

> NumPy’s main namespace as well as the [numpy.fft](https://numpy.org/devdocs/reference/routines.fft.html#module-numpy.fft) and [numpy.linalg](https://numpy.org/devdocs/reference/routines.linalg.html#module-numpy.linalg) namespaces are compatible [[1]](https://numpy.org/devdocs/reference/array_api.html#id3) with the [2022.12 version](https://data-apis.org/array-api/2022.12/index.html) of the Python array API standard.
>
> NumPy aims to implement support for the [2023.12 version](https://data-apis.org/array-api/2023.12/index.html) and future versions of the standard - assuming that those future versions can be upgraded to given NumPy’s [backwards compatibility policy](https://numpy.org/neps/nep-0023-backwards-compatibility.html).
>
> For usage guidelines for downstream libraries and end users who want to write code that will work with both NumPy and other array libraries, we refer to the documentation of the array API standard itself and to code and developer-focused documention in SciPy and scikit-learn.

But it's surely not the case in 1.26 and before and we still want to support 1.26 for a few years, so at least in the short term we need to be able to special-case numpy.

Furthermore, even once we drop support for numpy versions that are not compliant, we might still want to **occasionally** special case numpy, for instance to be able optimize memory usage by leveraging the `copy=False` kwarg of many numpy function that is not and will probably never be part of the Array API spec.

Reply:
I tried to remove two occurrences of `and not _is_numpy_namespace(xp)` that I believe should not be necessary and the tests seem to pass. Let me push this commit to the PR to run all the tests.

Reply:
Done in 09af4e3e33e3da8504e566edfbb1ce346a4cd410. Let's see if the CI likes it.

Reply:
It should work. At least when I quickly tested it locally it worked. My worry/though was not about this PR but about in the future when we use `train_test_split` or `_array_indexing` in a context where we do special case Numpy. Because if you remove the specia lcase in only one of the two it breaks. But maybe I'm worrying about a YAGNI

Reply:
I think I prefer the simpler version for now (no special casing). We can always decide to reintroduce it later if we face a situation that actually justifies the extra complexity.

Reply:
> @betatim About the `AttributeError: 'numpy.ndarray' object has no attribute '_array'`, isn't that surfacing a potential bug in `array_api_compat.numpy`? I think we should investigate and report a minimal reproducer upstream if that is the case.

To reproduce it:
```python
import numpy.array_api as xp
import numpy as np

X = xp.reshape(xp.arange(100), (10, 10))
idx = np.array([9, 1, 6, 7, 3, 0, 5])
xp.take(X, idx, axis=0)
```
I am not sure if this is a bug or a feature. I can see the argument for "don't mix Array API and not Array API" - we are using a Numpy array to index a Array API array.

The reason we end up with "Numpy indexing array api" is that the `if` clauses don't match up. As long as they are in sync everything is fine.